### PR TITLE
fix(core): harden local-ydb lifecycle readiness

### PR DESCRIPTION
**Pull request by astandrik**
<h3>Greptile Summary</h3>

This PR hardens the local-ydb startup lifecycle by replacing one-shot `ydbCli` calls with a new `waitForCommand`/`waitForYdbCli` retry loop that handles transient gRPC readiness errors, adds idempotent Docker volume removal via `removeVolumeIfPresentSpec`, routes `inspectScheme` to the correct static or dynamic endpoint based on path prefix, and widens the teardown skip-condition to cover connectivity failures in addition to auth failures.

- `waitForCommand` is a new exported shell-script generator that retries a command against a configurable error pattern, preserving the last exit code after exhaustion, and is exercised by three new unit tests covering success, fast-fail, and exhaustion scenarios.
- `cleanupStorage` replaces bare `docker volume rm` with an inspect-then-rm script that handles TOCTOU \"no such volume\" races idempotently while still surfacing genuine Docker failures.
- `inspectScheme` now routes paths under `rootDatabase` to `ydbRootCli` (static gRPC port) and paths under `tenantPath` to `ydbCli` (dynamic gRPC port), with the routing logic covered by two new scheme tests.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all changed paths are additive lifecycle-readiness hardening with good test coverage.

The retry loop, idempotent volume removal, and scheme routing logic are all well-contained and backed by new unit tests. The two observations are cosmetic edge cases that do not affect any current caller.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| packages/core/src/operations/commands.ts | Introduces `waitForCommand` and `waitForYdbCli`; refactors `ydbCli`/`ydbRootCli` to delegate to extracted `commandForYdb*` helpers. Minor: `rc` can be unbound if `maxAttempts` is 0 with `set -u` active. |
| packages/core/src/operations/scheme.ts | Adds `usesRootDatabase` routing logic; correctly prefers the tenant check before the root-database check to handle overlapping prefixes. |
| packages/core/src/operations/storage.ts | Replaces bare `docker volume rm` with idempotent `removeVolumeIfPresentSpec`; switches tenant-metadata check to `waitForYdbCli`. |
| packages/core/src/operations/stack.ts | Replaces one-shot `ydbCli` with `waitForYdbCli` in bootstrap; widens teardown skip-condition to network connectivity errors and `No such container`. |
| packages/core/src/operations/auth-operations.ts | Adds post-auth-hardening `waitForYdbCli` readiness probe; improves `setRootPassword` diagnostics by capturing command output to `last_error` instead of `/dev/null`. |
| packages/core/test/commands.test.ts | New test file covering retry-success, fast-fail, and exhaustion paths for `waitForCommand`. |
| packages/core/test/operations.test.ts | Adds tests for connectivity-based teardown skip, `removeVolumeIfPresentSpec`, and `last_error` temp-file allocation in `setRootPassword`. |
| packages/core/test/scheme.test.ts | Adds tests verifying root-database endpoint routing for root paths and non-tenant children. |

</details>

<a href="https://chatgpt.com/codex/deeplink?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22astandrik%2Flocal-ydb-toolkit%22%20on%20the%20existing%20branch%20%22codex%2Ffix-local-ydb-mcp-lifecycle-readiness%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22codex%2Ffix-local-ydb-mcp-lifecycle-readiness%22.%0A%0AFix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Apackages%2Fcore%2Fsrc%2Foperations%2Fcommands.ts%3A281-288%0AIf%20%60maxAttempts%60%20is%20passed%20as%20%600%60%2C%20%60seq%201%200%60%20produces%20no%20output%20and%20the%20loop%20body%20never%20executes%2C%20leaving%20%60rc%60%20unbound.%20With%20%60set%20-u%60%20active%2C%20%60exit%20%22%24rc%22%60%20then%20aborts%20with%20an%20%22unbound%20variable%22%20error%20instead%20of%20a%20clean%20exit.%20Initializing%20%60rc%60%20before%20the%20loop%20makes%20the%20edge%20case%20safe%20and%20also%20makes%20the%20intent%20clearer.%0A%0A%60%60%60suggestion%0A%20%20const%20maxAttempts%20%3D%20options.maxAttempts%20%3F%3F%2030%3B%0A%20%20const%20retryDelaySeconds%20%3D%20options.retryDelaySeconds%20%3F%3F%202%3B%0A%20%20return%20bash%28%5B%0A%20%20%20%20%22set%20-euo%20pipefail%22%2C%0A%20%20%20%20%22tmp%3D%24%28mktemp%29%22%2C%0A%20%20%20%20%22trap%20'rm%20-f%20%5C%22%24tmp%5C%22'%20EXIT%22%2C%0A%20%20%20%20%22rc%3D0%22%2C%0A%20%20%20%20%60for%20attempt%20in%20%24%28seq%201%20%24%7BmaxAttempts%7D%29%3B%20do%60%2C%0A%20%20%20%20%22%20%20rc%3D0%22%2C%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Apackages%2Fcore%2Fsrc%2Foperations%2Fauth-operations.ts%3A262%0A%60last_error%60%20is%20written%20with%20%60%3E%60%20%28truncate%29%2C%20so%20each%20%60rotate_with_password_file%60%20call%20silently%20overwrites%20the%20previous%20attempt's%20output.%20If%20the%20first%20attempt%20fails%20with%20the%20most%20informative%20diagnostic%20%28e.g.%2C%20a%20real%20authentication%20error%29%20and%20later%20attempts%20produce%20an%20empty%20or%20generic%20message%2C%20that%20original%20error%20is%20lost.%20Using%20%60%3E%3E%60%20%28append%29%20would%20preserve%20the%20full%20failure%20trail.%0A%0A%60%60%60suggestion%0A%20%20cat%20%22%24file%22%20%7C%20docker%20exec%20-i%20%24%7BshellQuote%28ctx.profile.staticContainer%29%7D%20bash%20-lc%20%24%7BshellQuote%28%60umask%20077%3B%20cat%20%3E%2Ftmp%2Froot.password%3B%20%2Fydb%20-e%20grpc%3A%2F%2Flocalhost%3A%24%7Bctx.profile.ports.dynamicGrpc%7D%20-d%20%24%7BshellQuote%28ctx.profile.tenantPath%29%7D%20--user%20%24%7BshellQuote%28ctx.profile.rootUser%29%7D%20--password-file%20%2Ftmp%2Froot.password%20yql%20-s%20%22ALTER%20USER%20%24%7Bctx.profile.rootUser%7D%20PASSWORD%20'%24%7BescapedPassword%7D'%3B%22%3B%20rc%3D%24%3F%3B%20rm%20-f%20%2Ftmp%2Froot.password%3B%20exit%20%24rc%60%29%7D%20%3E%3E%22%24last_error%22%202%3E%261%0A%60%60%60%0A%0A"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodexDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInCodex.svg?v=3">Fix All in Codex</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Apackages%2Fcore%2Fsrc%2Foperations%2Fcommands.ts%3A281-288%0AIf%20%60maxAttempts%60%20is%20passed%20as%20%600%60%2C%20%60seq%201%200%60%20produces%20no%20output%20and%20the%20loop%20body%20never%20executes%2C%20leaving%20%60rc%60%20unbound.%20With%20%60set%20-u%60%20active%2C%20%60exit%20%22%24rc%22%60%20then%20aborts%20with%20an%20%22unbound%20variable%22%20error%20instead%20of%20a%20clean%20exit.%20Initializing%20%60rc%60%20before%20the%20loop%20makes%20the%20edge%20case%20safe%20and%20also%20makes%20the%20intent%20clearer.%0A%0A%60%60%60suggestion%0A%20%20const%20maxAttempts%20%3D%20options.maxAttempts%20%3F%3F%2030%3B%0A%20%20const%20retryDelaySeconds%20%3D%20options.retryDelaySeconds%20%3F%3F%202%3B%0A%20%20return%20bash%28%5B%0A%20%20%20%20%22set%20-euo%20pipefail%22%2C%0A%20%20%20%20%22tmp%3D%24%28mktemp%29%22%2C%0A%20%20%20%20%22trap%20'rm%20-f%20%5C%22%24tmp%5C%22'%20EXIT%22%2C%0A%20%20%20%20%22rc%3D0%22%2C%0A%20%20%20%20%60for%20attempt%20in%20%24%28seq%201%20%24%7BmaxAttempts%7D%29%3B%20do%60%2C%0A%20%20%20%20%22%20%20rc%3D0%22%2C%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Apackages%2Fcore%2Fsrc%2Foperations%2Fauth-operations.ts%3A262%0A%60last_error%60%20is%20written%20with%20%60%3E%60%20%28truncate%29%2C%20so%20each%20%60rotate_with_password_file%60%20call%20silently%20overwrites%20the%20previous%20attempt's%20output.%20If%20the%20first%20attempt%20fails%20with%20the%20most%20informative%20diagnostic%20%28e.g.%2C%20a%20real%20authentication%20error%29%20and%20later%20attempts%20produce%20an%20empty%20or%20generic%20message%2C%20that%20original%20error%20is%20lost.%20Using%20%60%3E%3E%60%20%28append%29%20would%20preserve%20the%20full%20failure%20trail.%0A%0A%60%60%60suggestion%0A%20%20cat%20%22%24file%22%20%7C%20docker%20exec%20-i%20%24%7BshellQuote%28ctx.profile.staticContainer%29%7D%20bash%20-lc%20%24%7BshellQuote%28%60umask%20077%3B%20cat%20%3E%2Ftmp%2Froot.password%3B%20%2Fydb%20-e%20grpc%3A%2F%2Flocalhost%3A%24%7Bctx.profile.ports.dynamicGrpc%7D%20-d%20%24%7BshellQuote%28ctx.profile.tenantPath%29%7D%20--user%20%24%7BshellQuote%28ctx.profile.rootUser%29%7D%20--password-file%20%2Ftmp%2Froot.password%20yql%20-s%20%22ALTER%20USER%20%24%7Bctx.profile.rootUser%7D%20PASSWORD%20'%24%7BescapedPassword%7D'%3B%22%3B%20rc%3D%24%3F%3B%20rm%20-f%20%2Ftmp%2Froot.password%3B%20exit%20%24rc%60%29%7D%20%3E%3E%22%24last_error%22%202%3E%261%0A%60%60%60%0A%0A&repo=astandrik%2Flocal-ydb-toolkit&pr=54&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaudeDark.svg?v=3"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixAllInClaude.svg?v=3">Fix All in Claude Code</picture></a>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
packages/core/src/operations/commands.ts:281-288
If `maxAttempts` is passed as `0`, `seq 1 0` produces no output and the loop body never executes, leaving `rc` unbound. With `set -u` active, `exit "$rc"` then aborts with an "unbound variable" error instead of a clean exit. Initializing `rc` before the loop makes the edge case safe and also makes the intent clearer.

```suggestion
  const maxAttempts = options.maxAttempts ?? 30;
  const retryDelaySeconds = options.retryDelaySeconds ?? 2;
  return bash([
    "set -euo pipefail",
    "tmp=$(mktemp)",
    "trap 'rm -f \"$tmp\"' EXIT",
    "rc=0",
    `for attempt in $(seq 1 ${maxAttempts}); do`,
    "  rc=0",
```

### Issue 2 of 2
packages/core/src/operations/auth-operations.ts:262
`last_error` is written with `>` (truncate), so each `rotate_with_password_file` call silently overwrites the previous attempt's output. If the first attempt fails with the most informative diagnostic (e.g., a real authentication error) and later attempts produce an empty or generic message, that original error is lost. Using `>>` (append) would preserve the full failure trail.

```suggestion
  cat "$file" | docker exec -i ${shellQuote(ctx.profile.staticContainer)} bash -lc ${shellQuote(`umask 077; cat >/tmp/root.password; /ydb -e grpc://localhost:${ctx.profile.ports.dynamicGrpc} -d ${shellQuote(ctx.profile.tenantPath)} --user ${shellQuote(ctx.profile.rootUser)} --password-file /tmp/root.password yql -s "ALTER USER ${ctx.profile.rootUser} PASSWORD '${escapedPassword}';"; rc=$?; rm -f /tmp/root.password; exit $rc`)} >>"$last_error" 2>&1
```

`````

</details>

<sub>Reviews (2): Last reviewed commit: ["fix(core): address readiness and cleanup..."](https://github.com/astandrik/local-ydb-toolkit/commit/d645d17d3fa087c76d52c9285abd03e0eb7b6a9f) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=32600882)</sub>

### CODE REVIEW

File: packages/core/src/operations/commands.ts [hunk 1]
```diff
@@ -268,11 +272,55 @@ export function ydbCli(profile: ResolvedLocalYdbProfile, args: string[], databas
...
+    "  else",
+    "    cat \"$tmp\" >&2",
+    "    exit \"$rc\"",
+    "  fi",
+    "done",
+    "cat \"$tmp\" >&2",
+    "exit 1"
+  ].join("\n"), {
+    allowFailure: true,
+    timeoutMs: 120_000,
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/commands.ts]: Addressed locally in `d645d17`. The retry loop now preserves the last command exit code after retry exhaustion instead of collapsing it to `1`.

File: packages/core/src/operations/commands.ts [hunk 2]
```diff
@@ -268,11 +272,55 @@ export function ydbCli(profile: ResolvedLocalYdbProfile, args: string[], databas
   };
 }
 
+export function waitForYdbCli(profile: ResolvedLocalYdbProfile, args: string[], database: string, description: string, options: { root?: boolean } = {}): CommandSpec {
+  const command = options.root
+    ? commandForYdbRootCli(profile, args)
+    : commandForYdbCli(profile, args, database);
+  const retryableErrors = "CLIENT_UNAUTHENTICATED|SCHEME_ERROR|No database found|connection refused|Endpoint list is empty|Could not resolve redirected path|Failed to connect|TRANSPORT_UNAVAILABLE|UNAVAILABLE";
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/commands.ts]: Addressed locally in `d645d17`. I narrowed the retry match from a bare `UNAVAILABLE` substring to `Status:[[:space:]]*UNAVAILABLE`, which keeps the known CLI readiness output without matching unrelated `*_UNAVAILABLE` text.

File: packages/core/src/operations/commands.ts [hunk 3]
```diff
@@ -268,11 +272,55 @@ export function ydbCli(profile: ResolvedLocalYdbProfile, args: string[], databas
   };
 }
 
+export function waitForYdbCli(profile: ResolvedLocalYdbProfile, args: string[], database: string, description: string, options: { root?: boolean } = {}): CommandSpec {
+  const command = options.root
+    ? commandForYdbRootCli(profile, args)
+    : commandForYdbCli(profile, args, database);
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/commands.ts]: Addressed locally in `d645d17`. I removed the unused `options.root` branch and simplified `waitForYdbCli` to the only call pattern currently used by the codebase.

File: packages/core/src/operations/storage.ts [hunk 4]
```diff
@@ -292,7 +292,7 @@ export async function cleanupStorage(ctx: ToolkitContext, options: MutatingOptio
     assertSafeCleanupTarget(volume);
   }
   const specs = [
-    ...volumes.map((volume) => bash(`docker volume rm ${shellQuote(volume)}`)),
+    ...volumes.map((volume) => bash(`if docker volume inspect ${shellQuote(volume)} >/dev/null 2>&1; then docker volume rm ${shellQuote(volume)}; fi`)),
```
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/storage.ts]: Addressed locally in `d645d17`, but not by adding a blanket `|| true`. `cleanupStorage` now treats `no such volume` as an idempotent terminal state in both the inspect and rm paths, while still surfacing unexpected Docker or permission failures.

[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/storage.ts]: Addressed locally in `d645d17`. `cleanupStorage` now distinguishes a genuinely absent volume from inspect/rm failures instead of treating every inspect miss as success.

File: packages/core/src/operations/commands.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (astandrik) [reply packages/core/src/operations/commands.ts]: Addressed locally in `d645d17`. I added `packages/core/test/commands.test.ts` with `ShellCommandExecutor` coverage for retry-success, non-retryable fail-fast, and retry exhaustion preserving the final exit code.
